logical/join.go

```diff
@@ -0,0 +1,226 @@
+package logical
+
+import (
+	"context"
+	"runtime"
+
+	"github.com/cube2222/octosql"
+	"github.com/cube2222/octosql/config"
+	"github.com/cube2222/octosql/execution"
+	"github.com/cube2222/octosql/physical"
+	"github.com/cube2222/octosql/physical/metadata"
+	"github.com/pkg/errors"
+)
+
+type Join struct {
+	source   Node
+	joined   Node
+	joinType execution.JoinType // TODO: define this on every level, or import execution?
+	triggers []Trigger
+}
+
+func NewJoin(source, joined Node, joinType execution.JoinType) *Join {
+	return &Join{
+		source:   source,
+		joined:   joined,
+		joinType: joinType,
+	}
+}
+
+func (node *Join) Physical(ctx context.Context, physicalCreator *PhysicalPlanCreator) ([]physical.Node, octosql.Variables, error) {
+	var outNodes []physical.Node
+
+	sourceNodes, sourceVariables, err := node.source.Physical(ctx, physicalCreator)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get physical plan for join source nodes")
+	}
+
+	joinedNodes, joinedVariables, err := node.joined.Physical(ctx, physicalCreator)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get physical plan for left join joined nodes")
+	}
+
+	variables, err := sourceVariables.MergeWith(joinedVariables)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't merge variables for source and joined nodes")
+	}
+
+	// Based on the cardinality of sources we decide whether we will create a stream_join or a lookup_join
+	// Stream joins support only equity conjunctions (i.e a.x = b.y AND a.v + 17 = b.something * 2)
+	sourceCardinality := sourceNodes[0].Metadata().Cardinality()
+	joinedCardinality := joinedNodes[0].Metadata().Cardinality()
+
+	isStreamJoin := false
+
+	if sourceCardinality == metadata.Unbounded && joinedCardinality == metadata.Unbounded {
+		isStreamJoin = true
+	} else if joinedCardinality == metadata.BoundedFitsInLocalStorage {
+		isStreamJoin = true
+	} else if sourceCardinality == metadata.BoundedFitsInLocalStorage {
+		isStreamJoin = true
+	}
+
+	// Lookup join doesn't support outer join, so it must be a stream join
+	if node.joinType == execution.OUTER_JOIN {
+		isStreamJoin = true
+	}
+
+	if isStreamJoin {
+		var formula physical.Formula
+		if filter, ok := joinedNodes[0].(*physical.Filter); ok {
+			formula = filter.Formula
+			for i := range joinedNodes {
+				joinedNodes[i] = joinedNodes[i].(*physical.Filter).Source
+			}
+		} else {
+			formula = physical.NewConstant(true)
+		}
+
+		// We check if the ON part of the join is legal for a stream join
+		if !isConjunctionOfEqualities(formula) {
+			return nil, nil, errors.New("The ON part of join isn't a conjunction of equalities")
```

Comment:
If we have a join other than Unbounded X Unbounded, then we won't to fall back to a lookup join in this case.